File: bcc/two_edge.py
import sys
input = sys.stdin.readline 
sys.setrecursionlimit(1000000)

n, m = map(int, input().split())

tin = [0]*(n+1) 
low = [0]*(n+1)
g = [[] for _ in range(n+1)]
scc = 0
timer = 0
stack = []
two_edge_components = [[] for _ in range(n+1)]

for i in range(m):
    u, v = map(int, input().split())
    g[u].append(v)
    g[v].append(u)

def dfs(v, f): 
    global timer, scc 
    timer += 1
    tin[v] = timer
    low[v] = timer
    stack.append(v)
    mult_edges = False 
    for to in g[v]:
        if to == f and not mult_edges:
            mult_edges = True 
            continue
        if not tin[to]:
            dfs(to, v)
            low[v] = min(low[v], low[to])
        else:
            low[v] = min(low[v], tin[to])
    if low[v] == tin[v]:
        while stack[-1] != v:
            two_edge_components[scc].append(stack.pop())
        two_edge_components[scc].append(stack.pop())
        scc += 1

for i in range(n):
    if not tin[i]: dfs(i, -1)

print(scc)
for c in range(scc):
    print(len(two_edge_components[c]), end=' ')
    for v in two_edge_components[c]:
        print(v, end=' ')
    print()

File: bcc/two_edge.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <stack>
using namespace std;
const int maxn = 3e5;
int tin[maxn];
int low[maxn];
vector<int> g[maxn];
vector<int> two_edge_components[maxn];
int timer = 0;
int scc = 0;
stack<int> st;

void dfs(int v, int f) {
    tin[v] = low[v] = ++timer;
    bool mult_edges = false;
    st.push(v);
    for (auto to : g[v]) {
        if (to == f && !mult_edges) {
            mult_edges = true;
            continue;
        }
        if (!tin[to]) {
            dfs(to, v);
            low[v] = min(low[v], low[to]);
        } else {
            low[v] = min(low[v], tin[to]);
        }
    }
    if (low[v] == tin[v]) {
        while (st.top() != v) {
            two_edge_components[scc].push_back(st.top()); st.pop();
        }
        two_edge_components[scc++].push_back(st.top()); st.pop();
    }
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);
    int n, m; cin >> n >> m;
    for (int i = 0; i < m; ++i) {
        int x, y; cin >> x >> y;
        g[x].push_back(y);
        g[y].push_back(x);
    }
    for (int v = 0; v < n; ++v) {
        if (tin[v] == 0) dfs(v,-1);
    }
    cout << scc << '\n';
    for (int i = 0; i < scc; i++) {
        cout << two_edge_components[i].size() << ' ';
        for (int node : two_edge_components[i]) cout << node << ' ';
        cout << '\n';
    }
    return 0;
}